In [1]:
#import library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from tabulate import tabulate

In [2]:
netflix = pd.read_csv("netflix_titles.csv")
netflix.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [3]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [4]:
netflix.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

The dataset we are dealing with was obtained from Kaggle at https://www.kaggle.com/datasets/shivamb/netflix-shows. 
This dataset consists of 8807 rows and 12 features:

- show_id: Netflix ID of the media.
- type: Movie or TV Show.
- title: Title of the media.
- director: Director of the media.
- cast: Cast of the media.
- country: Country in which the movie was made.
- date_added: Date in which the media was added.
- release_year: Year in which the media was released.
- rating: Age rating of the media.
- duration: Duration of the media.
- listen_in: Classification given by Netflix.
- description: A short description of the media.

# Missing Data

In [5]:
netflix.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [6]:
#Show missing values in %

header = ["Column", "Missing Value (%)"]
missing_value_data = []
for col in netflix.columns:
    nulls = (netflix[col].isna().sum() / len(netflix) * 100)
    if nulls > 0 :
        missing_value_data.append([col, round(nulls,2)])
        #print("{} missing values: {}%".format(col, round(nulls,2)))
        
print(tabulate(missing_value_data, headers=header, tablefmt="fancy_grid"))

╒════════════╤═════════════════════╕
│ Column     │   Missing Value (%) │
╞════════════╪═════════════════════╡
│ director   │               29.91 │
├────────────┼─────────────────────┤
│ cast       │                9.37 │
├────────────┼─────────────────────┤
│ country    │                9.44 │
├────────────┼─────────────────────┤
│ date_added │                0.11 │
├────────────┼─────────────────────┤
│ rating     │                0.05 │
├────────────┼─────────────────────┤
│ duration   │                0.03 │
╘════════════╧═════════════════════╛


### Dealing with missing values

There are 6 features with missing values.
We must considers first three rows of the previous table, since they have more than 9% missing values. The others can be dropped.

"Director" and "Cast" will be replaced with "Data not available".

In [7]:
netflix['director'].replace(np.nan, 'Data not available', inplace = True)
netflix['cast'].replace(np.nan, 'Data not available', inplace = True)

In [8]:
netflix[netflix['country'].isnull()]

show_id     type                                title  \
2         s3  TV Show                            Ganglands   
3         s4  TV Show                Jailbirds New Orleans   
5         s6  TV Show                        Midnight Mass   
6         s7    Movie     My Little Pony: A New Generation   
10       s11  TV Show  Vendetta: Truth, Lies and The Mafia   
...      ...      ...                                  ...   
8718   s8719    Movie               Westside vs. the World   
8759   s8760    Movie               World's Weirdest Homes   
8783   s8784  TV Show                                 Yoko   
8785   s8786  TV Show                                  YOM   
8803   s8804  TV Show                          Zombie Dumb   

                           director  \
2                   Julien Leclercq   
3                Data not available   
5                     Mike Flanagan   
6     Robert Cullen, José Luis Ucha   
10               Data not available   
...                             ...   
8718                  Michael Fahey   
8759               Storm Theunissen   
8783             Data not available   
8785             Data not available   
8803             Data not available   

                                                   cast country  \
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...     NaN   
3                                    Data not available     NaN   
5     Kate Siegel, Zach Gilford, Hamish Linklater, H...     NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...     NaN   
10                                   Data not available     NaN   
...                                                 ...     ...   
8718                         Ron Perlman, Louie Simmons     NaN   
8759                                     Charlie Luxton     NaN   
8783  Eileen Stevens, Alyson Leigh Rosenfeld, Sarah ...     NaN   
8785  Sairaj, Devyani Dagaonkar, Ketan Singh, Mayur ...     NaN   
8803                                 Data not available     NaN   

              date_added  release_year rating   duration  \
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
5     September 24, 2021          2021  TV-MA   1 Season   
6     September 24, 2021          2021     PG     91 min   
10    September 24, 2021          2021  TV-MA   1 Season   
...                  ...           ...    ...        ...   
8718      August 9, 2019          2019  TV-MA     96 min   
8759    February 1, 2019          2015  TV-PG     49 min   
8783       June 23, 2018          2016   TV-Y   1 Season   
8785        June 7, 2018          2016  TV-Y7   1 Season   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   

                                              listed_in  \
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
5                    TV Dramas, TV Horror, TV Mysteries   
6                              Children & Family Movies   
10    Crime TV Shows, Docuseries, International TV S...   
...                                                 ...   
8718                       Documentaries, Sports Movies   
8759                                             Movies   
8783                                           Kids' TV   
8785                                           Kids' TV   
8803             Kids' TV, Korean TV Shows, TV Comedies   

                                            description  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
5     The arrival of a charismatic young priest brin...  
6     Equestria's divided. But a bright-eyed hero be...  
10    Sicily boasts a bold "Anti-Mafia" coalition. B...  
...                                                 ...  
8718  A look into the journey of influential strengt...  
8759  From a bubble-shaped palace to an island built...  
8783  Friends Mai, Oto and Vik's g

I will replace "country" missing values with the most common country (instead of dropping all of them)

In [9]:
netflix['country'] = netflix['country'].fillna(netflix['country'].mode()[0])

In [10]:
# Drops all the others
netflix.dropna(inplace = True)

In [11]:
netflix.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

# Duplicate

In [12]:
# Checking if there are any duplicates
netflix.duplicated().value_counts()

False    8790
dtype: int64

In [13]:
# Drop Duplicates
netflix.drop_duplicates(inplace= True)

In [14]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8790 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8790 non-null   object
 1   type          8790 non-null   object
 2   title         8790 non-null   object
 3   director      8790 non-null   object
 4   cast          8790 non-null   object
 5   country       8790 non-null   object
 6   date_added    8790 non-null   object
 7   release_year  8790 non-null   int64 
 8   rating        8790 non-null   object
 9   duration      8790 non-null   object
 10  listed_in     8790 non-null   object
 11  description   8790 non-null   object
dtypes: int64(1), object(11)
memory usage: 892.7+ KB


Now the dataset as been reduced to 8790 rows.

In [15]:
netflix.head(3)

show_id     type                 title            director  \
0      s1    Movie  Dick Johnson Is Dead     Kirsten Johnson   
1      s2  TV Show         Blood & Water  Data not available   
2      s3  TV Show             Ganglands     Julien Leclercq   

                                                cast        country  \
0                                 Data not available  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...  United States   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...

# Data Transformation

#### "date_added"

In [16]:
netflix["date_added"] = pd.to_datetime(netflix['date_added'])

In [17]:
netflix['month_added']=netflix['date_added'].dt.month
netflix['month_name_added']=netflix['date_added'].dt.month_name()
netflix['year_added'] = netflix['date_added'].dt.year

In [18]:
netflix.head(3)

show_id     type                 title            director  \
0      s1    Movie  Dick Johnson Is Dead     Kirsten Johnson   
1      s2  TV Show         Blood & Water  Data not available   
2      s3  TV Show             Ganglands     Julien Leclercq   

                                                cast        country  \
0                                 Data not available  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...  United States   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  month_added  \
0  As her father nears the end of his life, filmm...            9   
1  After crossing paths at a party, a Cape Town t...            9   
2  To protect his family from a powerful drug lor...            9   

  month_name_added  year_added  
0        September        2021  
1        September        2021  
2        September        2021

#### "listed_in"

In [19]:
netflix["listed_in"].unique()

array(['Documentaries', 'International TV Shows, TV Dramas, TV Mysteries',
       'Crime TV Shows, International TV Shows, TV Action & Adventure',
       'Docuseries, Reality TV',
       'International TV Shows, Romantic TV Shows, TV Comedies',
       'TV Dramas, TV Horror, TV Mysteries', 'Children & Family Movies',
       'Dramas, Independent Movies, International Movies',
       'British TV Shows, Reality TV', 'Comedies, Dramas',
       'Crime TV Shows, Docuseries, International TV Shows',
       'Dramas, International Movies',
       'Children & Family Movies, Comedies',
       'British TV Shows, Crime TV Shows, Docuseries',
       'TV Comedies, TV Dramas', 'Documentaries, International Movies',
       'Crime TV Shows, Spanish-Language TV Shows, TV Dramas',
       'Thrillers',
       'International TV Shows, Spanish-Language TV Shows, TV Action & Adventure',
       'International TV Shows, TV Action & Adventure, TV Dramas',
       'Comedies, International Movies',
       'Comedies, 

I'll take only the first 3 categories.

In [20]:
netflix.shape[0]

8790

In [21]:
netflix['listed_in_1'] = 0
netflix['listed_in_2'] = 0
netflix['listed_in_3'] = 0

tmp = netflix['listed_in'].str.split(',')

for i, row in netflix.iterrows():
    t = len(tmp[i])
    if t == 1:
        netflix['listed_in_1'][i] = tmp[i][0]
        netflix['listed_in_2'][i] = 0
        netflix['listed_in_3'][i] = 0
    if t == 2:
        netflix['listed_in_1'][i] = tmp[i][0]
        netflix['listed_in_2'][i] = tmp[i][1]
        netflix['listed_in_3'][i] = 0
    if t == 3:
        netflix['listed_in_1'][i] = tmp[i][0]
        netflix['listed_in_2'][i] = tmp[i][1]
        netflix['listed_in_3'][i] = tmp[i][2]

<ipython-input-21-deb8f26c619e>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  netflix['listed_in_1'][i] = tmp[i][0]
c:\python38\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)
<ipython-input-21-deb8f26c619e>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  netflix['listed_in_2'][i] = 0
<ipython-input-21-deb8f26c619e>:12: SettingWithCopyWarning:

In [22]:
netflix.head(3)

show_id     type                 title            director  \
0      s1    Movie  Dick Johnson Is Dead     Kirsten Johnson   
1      s2  TV Show         Blood & Water  Data not available   
2      s3  TV Show             Ganglands     Julien Leclercq   

                                                cast        country  \
0                                 Data not available  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...  United States   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   

                                         description  month_added  \
0  As her father nears the end of his life, filmm...            9   
1  After crossing paths at a party, a Cape Town t...            9   
2  To protect his family from a powerful drug lor...            9   

  month_name_added  year_added             listed_in_1  \
0        September        2021           Documentaries   
1        September        2021  International TV Shows   
2        September        2021          Crime TV Shows   

               listed_in_2             listed_in_3  
0                        0                       0  
1                TV Dramas            TV Mysteries  
2   International TV Shows   TV Action & Adventure

#### "type"

In [23]:
netflix["type"].unique()

array(['Movie', 'TV Show'], dtype=object)

Split dataset by type

In [24]:
tv_show_netflix = netflix[netflix["type"] == "TV Show"]
movie_netflix = netflix[netflix["type"] == "Movie"]

In [25]:
#alternativa allo split del dataset (creo una sola variabile in più)

#netflix['season_count'] = netflix.apply(lambda x : x['duration'].split(" ")[0] if "Season" in x['duration'] else "", axis = 1)
#netflix['duration'] = netflix.apply(lambda x : x['duration'].split(" ")[0] if "Season" not in x['duration'] else "", axis = 1)
#netflix.head()

#### "duration"

In [26]:
tv_show_netflix["duration"].unique()

array(['2 Seasons', '1 Season', '9 Seasons', '4 Seasons', '5 Seasons',
       '3 Seasons', '6 Seasons', '7 Seasons', '10 Seasons', '8 Seasons',
       '17 Seasons', '13 Seasons', '15 Seasons', '12 Seasons',
       '11 Seasons'], dtype=object)

In [27]:
tv_show_netflix['duration_seasons'] = pd.to_numeric(tv_show_netflix['duration'].str.split(' ',expand=True)[0])

<ipython-input-27-610e6f73f0ab>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tv_show_netflix['duration_seasons'] = pd.to_numeric(tv_show_netflix['duration'].str.split(' ',expand=True)[0])


In [28]:
tv_show_netflix.head(3)

show_id     type                  title            director  \
1      s2  TV Show          Blood & Water  Data not available   
2      s3  TV Show              Ganglands     Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans  Data not available   

                                                cast        country  \
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...  United States   
3                                 Data not available  United States   

  date_added  release_year rating   duration  \
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   

                                           listed_in  \
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   

                                         description  month_added  \
1  After crossing paths at a party, a Cape Town t...            9   
2  To protect his family from a powerful drug lor...            9   
3  Feuds, flirtations and toilet talk go down amo...            9   

  month_name_added  year_added             listed_in_1  \
1        September        2021  International TV Shows   
2        September        2021          Crime TV Shows   
3        September        2021              Docuseries   

               listed_in_2             listed_in_3  duration_seasons  
1                TV Dramas            TV Mysteries                 2  
2   International TV Shows   TV Action & Adventure                 1  
3               Reality TV                       0                 1

In [29]:
movie_netflix["duration"].unique()

array(['90 min', '91 min', '125 min', '104 min', '127 min', '67 min',
       '94 min', '161 min', '61 min', '166 min', '147 min', '103 min',
       '97 min', '106 min', '111 min', '110 min', '105 min', '96 min',
       '124 min', '116 min', '98 min', '23 min', '115 min', '122 min',
       '99 min', '88 min', '100 min', '102 min', '93 min', '95 min',
       '85 min', '83 min', '113 min', '13 min', '182 min', '48 min',
       '145 min', '87 min', '92 min', '80 min', '117 min', '128 min',
       '119 min', '143 min', '114 min', '118 min', '108 min', '63 min',
       '121 min', '142 min', '154 min', '120 min', '82 min', '109 min',
       '101 min', '86 min', '229 min', '76 min', '89 min', '156 min',
       '112 min', '107 min', '129 min', '135 min', '136 min', '165 min',
       '150 min', '133 min', '70 min', '84 min', '140 min', '78 min',
       '64 min', '59 min', '139 min', '69 min', '148 min', '189 min',
       '141 min', '130 min', '138 min', '81 min', '132 min', '123 min',
       '65

In [30]:
movie_netflix['duration_minutes'] = pd.to_numeric(movie_netflix['duration'].str.split(' ',expand=True)[0])

<ipython-input-30-d742c6e26a38>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movie_netflix['duration_minutes'] = pd.to_numeric(movie_netflix['duration'].str.split(' ',expand=True)[0])


In [31]:
movie_netflix.head(3)

show_id   type                             title  \
0      s1  Movie              Dick Johnson Is Dead   
6      s7  Movie  My Little Pony: A New Generation   
7      s8  Movie                           Sankofa   

                        director  \
0                Kirsten Johnson   
6  Robert Cullen, José Luis Ucha   
7                   Haile Gerima   

                                                cast  \
0                                 Data not available   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   

                                             country date_added  release_year  \
0                                      United States 2021-09-25          2020   
6                                      United States 2021-09-24          2021   
7  United States, Ghana, Burkina Faso, United Kin... 2021-09-24          1993   

  rating duration                                         listed_in  \
0  PG-13   90 min                                     Documentaries   
6     PG   91 min                          Children & Family Movies   
7  TV-MA  125 min  Dramas, Independent Movies, International Movies   

                                         description  month_added  \
0  As her father nears the end of his life, filmm...            9   
6  Equestria's divided. But a bright-eyed hero be...            9   
7  On a photo shoot in Ghana, an American model s...            9   

  month_name_added  year_added               listed_in_1          listed_in_2  \
0        September        2021             Documentaries                    0   
6        September        2021  Children & Family Movies                    0   
7        September        2021                    Dramas   Independent Movies   

             listed_in_3  duration_minutes  
0                      0                90  
6                      0                91  
7   International Movies               125

# Save to csv 3 dataset

In [32]:
netflix.to_csv('Netflix_cleaned.csv')
tv_show_netflix.to_csv('Netflix_TV_cleaned.csv')
movie_netflix.to_csv('Netflix_Movie_cleaned.csv')